### DISCUSSION TITLE
Sampling from a depth texture

### DISCUSSION THREAD

**Discussion by jwwalker** [google/filament#3831] [Q&A]
I'm experimenting with a material whose fragment shader samples from a depth texture associated with a `RenderTarget`, and I'm wondering about a couple things.  First, is this a legitimate thing to do?  I'm new to GLSL_ES, and the language specification says some things that I don't really understand but sound scary, such as

> If a non-shadow texture call is made to a sampler that represents a depth texture with depth comparisons turned on, then results are undefined.

Second, if the concept is OK, I was expecting the far plane to have depth 1, but experimentally it appears to have depth 0.  Why?

**Comment by romainguy** [comment]
You can sample from a depth texture you rendered into, that's how shadow maps, ambient occlusion, etc. work internally.

The part of the spec you are quoting means that if you use `texture()` to sample from a depth texture, you'll get an undefined result if said texture has a depth comparison mode set (see `TextureSampler::SamplerCompareMode`). It should be fine unless you've set a compare mode on your texture sampler. If you have set a compare mode, you should use a shadow sampler and provide a Z value so the texture call returns 0 or 1 as the result of the comparison.

Where do you get a depth of 1? Note that we use reverse Z when we can to improve depth buffer precision.

**Reply by jwwalker** [reply]
When I said that I expected the far plane to have depth 1, I meant going by previous OpenGL experience (though it's certainly possible that I'm just confused).  If you sometimes use reverse Z and sometimes not, how do I know how to interpret values obtained by sampling a depth texture?
